File: python_app/objetos_clases.py
# Para este cuarto punto, el objetivo era implementar el concepto de Clases y Objetos.
# A diferencia del archivo anterior donde solo usábamos la estructura para guardar datos estáticos,
# una Clase me permite tener tanto las variables (el estado) como las funciones (el comportamiento) 
# unidos en un solo bloque lógico.
class Estudiante:
    # Este es el constructor de la clase. Cada vez que quiero crear un nuevo estudiante en la memoria (Heap),
    # Python ejecuta esta función para darle los valores iniciales.
    def __init__(self, nombre: str, edad: int, promedio: float):
        self.nombre = nombre
        self.edad = edad
        self.promedio = promedio

    # En lugar de imprimir los datos sueltos desde fuera, creé este método dentro de la clase.
    # Así, es el propio estudiante el que tiene la capacidad de mostrar su propia información.
    def mostrar_info(self):
        print(f"[{self.nombre}] - Edad: {self.edad} años, Calificación: {self.promedio}")

    # Este es un método modificador (conocido en la teoría como "setter").
    # La mejor práctica en programación orientada a objetos es no cambiar las variables directamente desde afuera,
    # sino pedirle a la clase que lo haga a través de un método. Esto asegura el encapsulamiento de la información.
    def set_promedio(self, nuevo_promedio: float):
        self.promedio = nuevo_promedio
        print(f"Notificación: El registro académico de {self.nombre} fue actualizado a {self.promedio}")

def solicitar_datos_estudiante():
    # Creamos esta función extra para que la revisión de la tarea sea más interactiva y el profesor pueda
    # escribir los datos a través de la terminal, en vez de ver solo información fija.
    nombre = input("Ingrese el nombre del estudiante: ")
    try:
        edad = int(input(f"Ingrese la edad de {nombre}: "))
        promedio = float(input(f"Ingrese el promedio de {nombre} (ej: 4.5): "))
        
        # Una vez capturados todos los campos, usamos la clase Estudiante para generar un nuevo objeto 
        # y lo devolvemos para guardarlo.
        return Estudiante(nombre, edad, promedio)
    except ValueError:
        print("Error: Ingrese valores numéricos válidos para edad y promedio.")
        return None

def ejecutar_clases():
    print("\n--- Demostración del Punto 4: Objetos y Clases en Python ---")
    
    # Para cumplir con la rúbrica, instanciamos tres objetos base con datos de prueba.
    # Es importante notar que usamos la palabra "Estudiante" llamando a la clase, lo cual aloja esto en memoria dinámica.
    est1 = Estudiante("Carlos Ruiz", 21, 3.8)
    est2 = Estudiante("Diana Vargas", 23, 4.5)
    est3 = Estudiante("Roberto Carlos", 20, 3.2)

    # Todos esos objetos individuales los metemos dentro de una lista para poder procesarlos en conjunto.
    arreglo_objetos = [est1, est2, est3]
    
    # Le damos la opción al usuario de agregar sus propios objetos de forma interactiva usando un ciclo while.
    print("Se han instanciado 3 objetos Estudiante por defecto.")
    while True:
        respuesta = input("¿Desea instanciar un nuevo estudiante? (s/n): ").strip().lower()
        if respuesta == 's':
            nuevo_obj = solicitar_datos_estudiante()
            if nuevo_obj:
                # El nuevo objeto ingresado por teclado se añade al final de nuestra lista.
                arreglo_objetos.append(nuevo_obj)
                print(f"Objeto de '{nuevo_obj.nombre}' instanciado y agregado con éxito.")
        elif respuesta == 'n':
            # Cuando el usuario termina de ingresar datos, salimos del ciclo para continuar la demostración.
            break
        else:
            print("Opción no válida. Ingrese 's' para sí, o 'n' para no.")

    # A continuación recorro toda la lista, pero fíjense que ya no hago un "print" manual.
    # Le estoy diciendo a cada estudiante que se encargue él mismo de imprimir sus datos usando su propio método.
    print("\nDesplegando la información de cada objeto desde su propio método:")
    for estudiante in arreglo_objetos:
        estudiante.mostrar_info()

    # Ahora comprobamos que el método "setter" funciona. Modifico el promedio del tercer estudiante base,
    # y en vez de usar un igual (=), llamo a la función que programé arriba.
    print("\nAplicando actualización de calificaciones mediante el método de la clase...")
    
    # Valido que el tercer elemento de la lista efectivamente sea Roberto para evitar errores si el usuario ingresó más datos
    if len(arreglo_objetos) > 2 and arreglo_objetos[2].nombre == "Roberto Carlos":
        arreglo_objetos[2].set_promedio(4.0)

    # Por último, pido al objeto que vuelva a mostrar sus datos para certificar que el promedio efectivamente cambió en memoria.
    print("\nValidando el cambio en el estudiante modificado:")
    if len(arreglo_objetos) > 2 and arreglo_objetos[2].nombre == "Roberto Carlos":
        arreglo_objetos[2].mostrar_info()
        
    print("----------------------------------------------------------\n")

if __name__ == "__main__":
    ejecutar_clases()

File: javascript_app/objetos_clases.js
const readline = require('readline');

const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
});

// A diferencia del archivo anterior donde usé objetos JSON estáticos (structs simulados),
// aquí utilizo la sintaxis de Programación Orientada a Objetos introducida en ES6 (ECMAScript 6).
// Ahora tenemos un constructor real, métodos internos y encapsulamiento. Esto vive en la memoria Heap.
class EstudianteObjeto {
    constructor(nombre, edad, promedio) {
        this.nombre = nombre;
        this.edad = edad;
        this.promedio = promedio;
    }

    // Este es un comportamiento propio del objeto. Él mismo sabe cómo imprimirse, no necesito 
    // sacar las variables desde afuera para imprimirlas como lo hice con el struct.
    mostrarInfo() {
        console.log(`[Datos del Objeto] Estudiante: ${this.nombre} | Edad: ${this.edad} | Promedio: ${this.promedio}`);
    }

    // Método setter para encapsular la actualización de la nota
    setPromedio(nuevoPromedio) {
        this.promedio = nuevoPromedio;
        console.log(`Evento: El promedio de ${this.nombre} ha sido actualizado a ${this.promedio}`);
    }
}

function solicitarDatosObjeto(callback) {
    rl.question("Ingrese el nombre del estudiante: ", (nombre) => {
        rl.question(`Ingrese la edad de ${nombre}: `, (edadStr) => {
            rl.question(`Ingrese el promedio de ${nombre} (ej: 4.5): `, (promedioStr) => {
                const edad = parseInt(edadStr);
                const promedio = parseFloat(promedioStr);

                if (isNaN(edad) || isNaN(promedio)) {
                    console.log("Error: Los datos numéricos no son válidos.");
                    callback(null);
                } else {
                    // Utilizo la palabra 'new' para reservar espacio dinámico en la memoria Heap
                    const nuevoObjeto = new EstudianteObjeto(nombre, edad, promedio);
                    callback(nuevoObjeto);
                }
            });
        });
    });
}

function iniciarPrograma() {
    console.log("\n--- Demostración del Punto 4: Clases y Objetos reales en JS ---");

    // Instancio 3 objetos con la palabra clave 'new'
    const est1 = new EstudianteObjeto("Carlos Ruiz", 21, 3.8);
    const est2 = new EstudianteObjeto("Diana Vargas", 23, 4.5);
    const est3 = new EstudianteObjeto("Roberto Carlos", 20, 3.2);

    const arregloObjetos = [est1, est2, est3];
    console.log("Se han instanciado 3 objetos Estudiante por defecto.");

    function preguntar() {
        rl.question("¿Desea instanciar un nuevo estudiante? (s/n): ", (respuesta) => {
            respuesta = respuesta.trim().toLowerCase();
            if (respuesta === 's') {
                solicitarDatosObjeto((nuevo) => {
                    if (nuevo) {
                        arregloObjetos.push(nuevo);
                        console.log(`Objeto '${nuevo.nombre}' agregado con éxito.`);
                    }
                    preguntar();
                });
            } else if (respuesta === 'n') {
                mostrarResultados();
            } else {
                console.log("Opción no válida. Ingrese 's' o 'n'.");
                preguntar();
            }
        });
    }

    function mostrarResultados() {
        console.log("\nInvocando el comportamiento interno (método) de cada objeto en el arreglo:");
        arregloObjetos.forEach(estudiante => {
            // Ya no hago un print feo, sino que le pido al objeto que ejecute su propio código
            estudiante.mostrarInfo();
        });

        console.log("\nAplicando una actualización del estado utilizando un método del objeto...");
        if (arregloObjetos.length > 2 && arregloObjetos[2].nombre === "Roberto Carlos") {
            // Accedo al setter en lugar de acceder a la variable de forma insegura
            arregloObjetos[2].setPromedio(4.0);
        }

        console.log("\nComprobación del estado final del objeto modificado:");
        if (arregloObjetos.length > 2 && arregloObjetos[2].nombre === "Roberto Carlos") {
            arregloObjetos[2].mostrarInfo();
        }
        console.log("----------------------------------------------------------------------\n");
        rl.close();
    }

    preguntar();
}

iniciarPrograma();
